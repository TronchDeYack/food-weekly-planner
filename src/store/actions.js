export const ADD_RECIPE_GET_INGREDIENTS = 'ADD_RECIPE_GET_INGREDIENTS';
