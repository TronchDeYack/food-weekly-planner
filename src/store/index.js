import Vue from 'vue';
import Vuex from 'vuex';

import AddRecipe from './modules/add-recipe';

Vue.use(Vuex);

export default new Vuex.Store({
  state: {
  },
  mutations: {
  },
  actions: {
  },
  modules: {
    AddRecipe,
  },
});
