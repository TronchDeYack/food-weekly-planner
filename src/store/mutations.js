export const ADD_RECIPE_NEXT_STEP = 'ADD_RECIPE_NEXT_STEP';
export const ADD_RECIPE_PREVIOUS_STEP = 'ADD_RECIPE_PREVIOUS_STEP';
export const ADD_RECIPE_SET_INGREDIENTS = 'ADD_RECIPE_SET_INGREDIENTS';
export const ADD_RECIPE_TOGGLE_INGREDIENT = 'ADD_RECIPE_TOGGLE_INGREDIENT';
