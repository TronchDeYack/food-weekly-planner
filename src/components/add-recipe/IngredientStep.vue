<template>
  <div>
    <StepTitle :step="step" :title="title" />
    <StepContent :step="step">
      <v-chip-group multiple active-class="primary">
        <v-chip label v-for="ingredient in filteredIngredients" :input-value="ingredient.selected"
          @click.native="toggleIngredient(ingredient)" :key="ingredient.id">
          {{ ingredient.label }}
        </v-chip>
      </v-chip-group>
    </StepContent>
  </div>
</template>

<script>
import StepTitle from '@/components/add-recipe/StepTitle.vue';
import StepContent from '@/components/add-recipe/StepContent.vue';

import { ADD_RECIPE_TOGGLE_INGREDIENT } from '@/store/mutations';

export default {
  name: 'add-recipe-ingredient-step',
  components: {
    StepContent,
    StepTitle,
  },
  props: {
    title: {
      type: String,
    },
    step: {
      type: Number,
    },
    ingredientType: {
      type: String,
    },
  },
  computed: {
    filteredIngredients() {
      return this.$store.getters.filteredIngredients(this.ingredientType);
    },
  },
  methods: {
    toggleIngredient(ingredient) {
      this.$store.commit(ADD_RECIPE_TOGGLE_INGREDIENT, {
        ingredientId: ingredient.id,
        selected: !ingredient.selected,
      });
    },
  },
};
</script>
