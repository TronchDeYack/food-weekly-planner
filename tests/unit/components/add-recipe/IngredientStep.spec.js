import { shallowMount } from '@vue/test-utils';

import IngredientStep from '@/components/add-recipe/IngredientStep.vue';
import { ADD_RECIPE_TOGGLE_INGREDIENT } from '@/store/mutations';

const ingredientType = 'CONDIMENT';

function getMockedStore() {
  return {
    commit: jest.fn(),
    state: {
      AddRecipe: {
        currentStep: 2,
        ingredients: [
          {
            id: 'PATE-ID',
            label: 'PÃ¢tes',
            unit: 'GR',
            type: 'STARCHY-FOOD',
            selected: true,
          },
          {
            id: 'OIL-ID',
            label: 'Huile d\'olives',
            unit: 'CAC',
            type: 'CONDIMENT',
            selected: false,
          },
          {
            id: 'TOMATE-ID',
            label: 'Tomates',
            unit: 'UNIT',
            type: 'VEGETABLE',
            selected: true,
          },
          {
            id: 'SALT-ID',
            label: 'Sel',
            unit: 'CAC',
            type: 'CONDIMENT',
            selected: true,
          },
        ],
      },
    },
    getters: {
      filteredIngredients: (type) => {
        if (type === ingredientType) {
          return [
            {
              id: 'SAUCE-TOMATE-ID',
              label: 'Sauce Tomate',
              unit: 'GR',
              type: 'CONDIMENT',
              selected: false,
            },
            {
              id: 'OIL-ID',
              label: 'Huile d\'olives',
              unit: 'CAC',
              type: 'CONDIMENT',
              selected: false,
            },
            {
              id: 'SALT-ID',
              label: 'Sel',
              unit: 'CAC',
              type: 'CONDIMENT',
              selected: true,
            },
          ];
        }
        return [];
      },
    },
  };
}

function createComponentInstance(propsData) {
  const $store = getMockedStore();
  return shallowMount(IngredientStep, {
    propsData,
    mocks: {
      $store,
    },
  });
}

describe('IngredientStep', () => {
  let wrapper;

  const propsData = {
    title: 'EAT ME',
    step: 2,
    ingredientType: 'CONDIMENT',
  };

  beforeEach(() => {
    wrapper = createComponentInstance(propsData);
  });

  describe('render', () => {
    it('Should render.', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('Should display as many chips as filtered ingredients.', () => {
      const filteredIngredients = [
        {
          id: 'SAUCE-TOMATE-ID',
          label: 'Sauce Tomate',
          unit: 'GR',
          type: 'CONDIMENT',
        },
        {
          id: 'OIL-ID',
          label: 'Huile d\'olives',
          unit: 'CAC',
          type: 'CONDIMENT',
        },
        {
          id: 'SALT-ID',
          label: 'Sel',
          unit: 'CAC',
          type: 'CONDIMENT',
        },
      ];

      expect(wrapper.findAll('v-chip-stub')).toHaveLength(filteredIngredients.length);
      filteredIngredients.forEach((ingredient, index) => {
        expect(wrapper.findAll('v-chip-stub').at(index).text()).toEqual(ingredient.label);
      });
    });
  });

  describe('methods', () => {
    describe('toggleIngredients', () => {
      it('Should select an ingredient by calling "ADD_RECIPE_TOGGLE_INGREDIENT" mutation with the selected ingredient ID when clicking on an unselected ingredient chip.', () => {
        wrapper.findAll('v-chip-stub').at(1).trigger('click');

        expect(wrapper.vm.$store.commit.mock.calls[0][0]).toEqual(ADD_RECIPE_TOGGLE_INGREDIENT, {
          ingredientId: 'OIL-ID',
          selected: true,
        });
      });

      it('Should unselect an ingredient by calling "ADD_RECIPE_TOGGLE_INGREDIENT" mutation with the selected ingredient ID when clicking on a selected ingredient chip.', () => {
        wrapper.findAll('v-chip-stub').at(2).trigger('click');

        expect(wrapper.vm.$store.commit.mock.calls[0][0]).toEqual(ADD_RECIPE_TOGGLE_INGREDIENT, {
          ingredientId: 'SALT-ID',
          selected: true,
        });
      });
    });
  });
});
