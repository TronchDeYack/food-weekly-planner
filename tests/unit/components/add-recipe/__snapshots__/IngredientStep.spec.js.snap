// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IngredientStep render Should render. 1`] = `
<div>
  <steptitle-stub title="EAT ME" step="2"></steptitle-stub>
  <stepcontent-stub step="2">
    <v-chip-group-stub activeclass="primary" multiple="true" nexticon="$next" mobilebreakpoint="1264" previcon="$prev">
      <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" label="true">
        Sauce Tomate
      </v-chip-stub>
      <v-chip-stub activeclass="" tag="span" active="true" closeicon="$delete" filtericon="$complete" label="true">
        Huile d'olives
      </v-chip-stub>
      <v-chip-stub activeclass="" tag="span" inputvalue="true" active="true" closeicon="$delete" filtericon="$complete" label="true">
        Sel
      </v-chip-stub>
    </v-chip-group-stub>
  </stepcontent-stub>
</div>
`;
